File: clients/scala/src/main/scala/ai_cup_22/debugging/Camera.scala
package ai_cup_22.debugging

import ai_cup_22.util.StreamUtil

/**
 * Camera state
 *
 * @param center Center
 * @param rotation Rotation
 * @param attack Attack angle
 * @param fov Vertical field of view
 */
case class Camera(center: ai_cup_22.model.Vec2, rotation: Double, attack: Double, fov: Double) {
    /**
     * Write Camera to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        center.writeTo(stream)
        StreamUtil.writeDouble(stream, rotation)
        StreamUtil.writeDouble(stream, attack)
        StreamUtil.writeDouble(stream, fov)
    }

    /**
     * Get string representation of Camera
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Camera { ")
        stringBuilder.append("center: ")
        stringBuilder.append(center)
        stringBuilder.append(", ")
        stringBuilder.append("rotation: ")
        stringBuilder.append(rotation)
        stringBuilder.append(", ")
        stringBuilder.append("attack: ")
        stringBuilder.append(attack)
        stringBuilder.append(", ")
        stringBuilder.append("fov: ")
        stringBuilder.append(fov)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Camera {
    /**
     * Read Camera from input stream
     */
    def readFrom(stream: java.io.InputStream): Camera = Camera(
        ai_cup_22.model.Vec2.readFrom(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream)
    )
}

File: clients/swift/Sources/AiCup22/Debugging/Camera.swift
/// Camera state
public struct Camera {
    /// Center
    let center: Vec2

    /// Rotation
    let rotation: Double

    /// Attack angle
    let attack: Double

    /// Vertical field of view
    let fov: Double

    /// Read Camera from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Camera {
        var center: Vec2
        center = Vec2.readFrom(stream)
        var rotation: Double
        rotation = stream.readDouble()
        var attack: Double
        attack = stream.readDouble()
        var fov: Double
        fov = stream.readDouble()
        return Camera(center: center, rotation: rotation, attack: attack, fov: fov)
    }

    /// Write Camera to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        center.writeTo(stream)
        stream.writeDouble(rotation)
        stream.writeDouble(attack)
        stream.writeDouble(fov)
    }
}
